zilliqa/src/consensus.rs

```diff
@@ -1579,25 +1607,49 @@ impl Consensus {
 
     // Checks for the validity of a block and adds it to our block store if valid.
     // Returns true when the block is valid and newly seen and false otherwise.
-    pub fn receive_block(&mut self, block: Block) -> Result<bool> {
+    pub fn receive_block(&mut self, proposal: Proposal) -> Result<bool> {
+        let (block, transactions) = proposal.into_parts();
+        trace!(
+            "received block: {} number: {}, view: {}",
+            block.hash(),
+            block.number(),
+            block.view()
+        );
         if self.block_store.contains_block(block.hash())? {
-            trace!("recieved block already seen: {}", block.hash());
+            trace!(
+                "recieved block already seen: {} - our head is {}",
+                block.hash(),
+                self.head_block()
+            );
+            return Ok(false);
+        }
+
+        // Check whether it is loose or not - we do not store loose blocks.
+        if !self.block_store.contains_block(block.parent_hash())? {
+            trace!("received block is loose: {}", block.hash());
+
+            warn!(
+                "missing received block the parent! Lets request the parent, then: {}",
+                block.parent_hash()
+            );
+            self.block_store
+                .request_blocks(None, block.header.number.saturating_sub(1))?;
```

Comment:
Likewise here: we've just received a block; we're requesting its parent now, but we are throwing away the block itself. Why not request up to the block's height, so if the parent does exist, we then store it _and_ the block?